lib/instrumentation/span.js

```diff
@@ -137,8 +212,86 @@ Span.prototype.setHttpContext = function (context) {
   this._http = Object.assign(this._http || {}, context)
 }
 
-Span.prototype.setDestinationContext = function (context) {
-  this._destination = Object.assign(this._destination || {}, context)
+/**
+ * This is deprecated and will be dropped in a future version. This was always
+ * an internal method, but possibly used by enterprising users of manual
+ * instrumentation.
+ *
+ * @deprecated Users should use the public `setServiceTarget()`.
+ *    Internal APM agent code should use `_setDestinationContext()`.
+ */
+Span.prototype.setDestinationContext = function (destCtx) {
+  process.emitWarning(
+    '<span>.setDestinationContext() is deprecated, use <span>.setServiceTarget().',
+    'DeprecationWarning',
+    'ELASTIC_APM_SET_DESTINATION_CONTEXT'
+  )
+
+  if (destCtx.service && destCtx.service.resource) {
+    this.setServiceTarget('', destCtx.service.resource)
+  }
+  const destCtxWithoutService = Object.assign({}, destCtx)
+  delete destCtxWithoutService.service
+  this._setDestinationContext(destCtxWithoutService)
+}
+
+/**
+ * The internal method for setting "destination" context.
+ *
+ * "destination.service.resource" should only ever be included for special
+ * cases. It is typically inferred from other fields via a general algorithm.
+ */
+Span.prototype._setDestinationContext = function (destCtx) {
+  this._destination = Object.assign(this._destination || {}, destCtx)
+}
```

Comment:
This is a longer term discussion that goes beyond this PR, but we don't seem to be super consistent on the semantic meaning of method names with a leading underscore. 

This usually means the method is meant to be analogous to a protected or private method in a language with access modifiers. 

However,  we're calling `_setDestinationContext` _as_ a public method in the instrumentation modules so in _this_ instance we seem to be saying that `_setDestinationContext` is a public method but just not a public method that's part of our API.

If we're going to adopt the leading underscore as a convention is would be good to get on the same page as to what it means and more generally how we think about enforcing API boundaries between our own modules.

Reply:
I responded a little bit below (https://github.com/elastic/apm-agent-nodejs/pull/2882#discussion_r971162895) as well.

Mostly this use of leading underscore to mean "internal to the agent" seemed inline with current common uses of the following in the instrumentation modules:

```
agent._instrumentation
agent._conf
span._setOutcomeFromHttpStatusCode
span._setOutcomeFromErrorCapture
```

However, documenting naming conventions would be good.
I'll bring up a separate discussion on conventions for this.